File: app/src/main/java/com/meteorhead/leave/models/helpers/impl/RealmWorkingDays.java
package com.meteorhead.leave.models.helpers.impl;

import com.meteorhead.leave.database.realm.HolidaysRealmService;
import com.meteorhead.leave.models.Holiday;
import com.meteorhead.leave.models.HolidayFields;
import com.meteorhead.leave.models.helpers.WorkingDays;
import com.meteorhead.leave.utils.DateConverter;
import com.orhanobut.logger.Logger;

import org.joda.time.DateTimeConstants;
import org.joda.time.Days;
import org.joda.time.LocalDate;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import javax.annotation.Nullable;

import io.realm.RealmResults;

/**
 * Created by wierzchanowskig on 25.09.2016.
 */
public class RealmWorkingDays implements WorkingDays {

    private LocalDate dateFrom;
    private LocalDate dateTo;
    private int daysCount = 0;
    private HolidaysRealmService holidaysDbService;
    private RealmResults<Holiday> holidayCache;

    public RealmWorkingDays(@Nullable Date from, @Nullable Date to) {
        if(from != null) {
            this.dateFrom = new LocalDate(from.getTime());
        }
        if(to != null) {
            this.dateTo = new LocalDate(to.getTime());
        }

        holidaysDbService = new HolidaysRealmService();
    }

    @Override
    public Date getLastWorkingDay(int days) {
        dateTo = dateFrom.plusDays(days-1);
        int workingDays = getWorkingDaysCount();
        int tempDays = workingDays;
        while (workingDays != days) {
            dateTo = dateFrom.plusDays(tempDays++);
            workingDays = getWorkingDaysCount();
        }
        return dateTo.toDate();
    }

    @Override
    public Date getFirstWorkingDay(int days) {
        dateTo = dateFrom.plusDays(0);
        dateFrom = dateTo.minusDays(days-1);
        int workingDays = getWorkingDaysCount();
        while (workingDays != days) {
            dateFrom = dateFrom.minusDays(1);
            workingDays = getWorkingDaysCount();
        }
        return dateFrom.toDate();
    }

    @Override
    public int getWorkingDaysCount() {
        return getAllDays() - getHolidaysCount() - getSaturdays() - getSundays();
    }

    @Override
    public int getHolidaysCount() {
        String countryCode = getDefaultCountryCode();
        List<Holiday> holidaysList = getHolidaysFromDb(dateFrom.toDate(), dateTo.toDate(), countryCode);
        int holidaysCount = 0;
        for (Holiday holiday :
                holidaysList) {
            LocalDate localDate = new LocalDate(holiday.getHolidayDate().getTime());
            if(localDate.getDayOfWeek() < 6){
                holidaysCount++;
            }
        }
        return holidaysCount;
    }

    @Override
    public int getAllDays() {
        daysCount = Days.daysBetween(dateFrom, dateTo).getDays() + 1;
        return daysCount;
    }

    @Override
    public int getSundays() {
        int sundays = daysCount / 7;
        int rest = daysCount % 7;
        if (dateFrom.getDayOfWeek() + rest > DateTimeConstants.SUNDAY) {
            sundays++;
        }
        return sundays;
    }

    @Override
    public int getSaturdays() {
        int saturdays = daysCount / 7;
        int rest = daysCount % 7;
        if (dateFrom.getDayOfWeek() + rest > DateTimeConstants.SATURDAY) {
            saturdays++;
        }
        return saturdays;
    }

    @Override
    public void setStartDate(Date startDate) {
        holidayCache = null;
        this.dateFrom = new LocalDate(startDate.getTime());
    }

    @Override
    public Calendar[] getAllFreeDays() {
        String countryCode = getDefaultCountryCode();
        List<Calendar> freeDaysList = new ArrayList<>();
        List<Holiday> holidays = getHolidaysFromDb(dateFrom.toDate(), dateTo.toDate(), countryCode);
        for (Holiday holiday : holidays) {
            Calendar holidayDate = DateConverter.getAsCalendar(holiday.getHolidayDate());
            freeDaysList.add(holidayDate);
        }

        LocalDate date = new LocalDate(dateFrom.toDate().getTime());
        int dayOfTheWeek = date.getDayOfWeek();
        int diffToSaturday = 6 - dayOfTheWeek;
        date = date.plusDays(diffToSaturday);
        while(date.toDate().getTime() < dateTo.toDate().getTime()) {
            freeDaysList.add(DateConverter.getAsCalendar(date.toDate()));
            date = date.plusDays(1);
            freeDaysList.add(DateConverter.getAsCalendar(date.toDate()));
            date = date.plusDays(6);
            Logger.i("loop " + date.toString());
        }

        return freeDaysList.toArray(new Calendar[0]);
    }

    private List<Holiday> getHolidaysFromDb(Date from, Date to, String countryCode) {
        if(holidayCache == null) {
            holidayCache = holidaysDbService.getHolidays(countryCode);
        }
        return holidayCache.where()
                .between(HolidayFields.HOLIDAY_DATE, from, to)
                .findAll();
    }

    private String getDefaultCountryCode() {
        return Locale.getDefault().getLanguage();
    }
}

File: app/src/main/java/com/meteorhead/leave/models/impl/RealmWorkingDays.kt
package com.meteorhead.leave.models.impl

import com.meteorhead.leave.data.Holiday
import com.meteorhead.leave.data.Leave
import com.meteorhead.leave.database.realm.HolidaysRealmService
import com.meteorhead.leave.models.WorkingDays
import com.meteorhead.leave.utils.DateConverter
import io.realm.Realm
import io.realm.RealmResults
import org.joda.time.DateTimeConstants
import org.joda.time.Days
import org.joda.time.LocalDate
import java.util.*

/**
 * Created by wierzchanowskig on 25.09.2016.
 */

class RealmWorkingDays(from: Date?, to: Date?) : WorkingDays {

    private var dateFrom: LocalDate? = null
    private var dateTo: LocalDate? = null
    private var daysCount = 0
    private val holidaysDbService: HolidaysRealmService = HolidaysRealmService(Realm.getDefaultInstance())
    private var holidayCache: RealmResults<Holiday>? = null

    init {
        if (from != null) {
            this.dateFrom = LocalDate(from.time)
        }
        if (to != null) {
            this.dateTo = LocalDate(to.time)
        }
    }

    override fun getLastWorkingDay(days: Int): Date {
        dateTo = dateFrom!!.plusDays(days - 1)
        var workingDays = workingDaysCount
        var tempDays = workingDays
        while (workingDays != days) {
            dateTo = dateFrom!!.plusDays(tempDays++)
            workingDays = workingDaysCount
        }
        return dateTo!!.toDate()
    }

    override fun getFirstWorkingDay(days: Int): Date {
        dateTo = dateFrom!!.plusDays(0)
        dateFrom = dateTo!!.minusDays(days - 1)
        var workingDays = workingDaysCount
        while (workingDays != days) {
            dateFrom = dateFrom!!.minusDays(1)
            workingDays = workingDaysCount
        }
        return dateFrom!!.toDate()
    }

    override fun getWorkingDaysCount(): Int {
        return allDays - holidaysCount - saturdays - sundays
    }

    override fun getHolidaysCount(): Int {
        val countryCode = defaultCountryCode
        val holidaysList = getHolidaysFromDb(dateFrom!!.toDate(), dateTo!!.toDate(), countryCode)
        val holidaysCount = holidaysList
            .map { LocalDate(it.holidayDate!!.time) }
            .count { it.dayOfWeek < 6 }
        return holidaysCount
    }

    override fun getAllDays(): Int {
        daysCount = Days.daysBetween(dateFrom!!, dateTo!!).days + 1
        return daysCount
    }

    override fun getSundays(): Int {
        var sundays = daysCount / 7
        val rest = daysCount % 7
        if (dateFrom!!.dayOfWeek + rest > DateTimeConstants.SUNDAY) {
            sundays++
        }
        return sundays
    }

    override fun getSaturdays(): Int {
        var saturdays = daysCount / 7
        val rest = daysCount % 7
        if (dateFrom!!.dayOfWeek + rest > DateTimeConstants.SATURDAY) {
            saturdays++
        }
        return saturdays
    }

    override fun setStartDate(startDate: Date) {
        holidayCache = null
        this.dateFrom = LocalDate(startDate.time)
    }

    override fun getAllFreeDays(): Array<Calendar> {
        val countryCode = defaultCountryCode
        val freeDaysList = ArrayList<Calendar>()
        val holidays = getHolidaysFromDb(dateFrom!!.toDate(), dateTo!!.toDate(), countryCode)
        for (holiday in holidays) {
            val holidayDate = DateConverter.getAsCalendar(holiday.holidayDate)
            freeDaysList.add(holidayDate)
        }

        var date = LocalDate(dateFrom!!.toDate().time)
        val dayOfTheWeek = date.dayOfWeek
        val diffToSaturday = 6 - dayOfTheWeek
        date = date.plusDays(diffToSaturday)
        while (date.toDate().time < dateTo!!.toDate().time) {
            freeDaysList.add(DateConverter.getAsCalendar(date.toDate()))
            date = date.plusDays(1)
            freeDaysList.add(DateConverter.getAsCalendar(date.toDate()))
            date = date.plusDays(6)
        }

        return freeDaysList.toTypedArray()
    }

    private fun getHolidaysFromDb(from: Date, to: Date, countryCode: String): List<Holiday> {
        if (holidayCache == null) {
            holidayCache = holidaysDbService.getHolidays(countryCode)
        }
        return holidayCache!!.where()
            .between(Holiday::holidayDate.name, from, to)
            .findAll()
    }

    private val defaultCountryCode: String
        get() = Locale.getDefault().language

    override fun getPropositions(from: LocalDate, to: LocalDate, days: Int): List<Leave> {
        var fromDate = from
        var toDate = to
        fromDate = fromDate.minusDays(days)
        toDate = toDate.plusDays(days)

        val holidays = holidaysDbService.getHolidaysBetweenDates(
            Locale.getDefault().isO3Country,
            fromDate.toDate(), toDate.toDate())

        val proposedLeave = ArrayList<Leave>()
        for (holiday in holidays) {
            var dayTo = LocalDate(holiday.holidayDate.time)
            if (dayTo.dayOfWeek > 6) {
                dayTo = dayTo.plusDays(6 - dayTo.dayOfWeek)
            }
            this.setStartDate(dayTo.toDate())
            proposedLeave.add(Leave(dayTo.toDate(), this.getLastWorkingDay(days), days))
            if (dayTo.dayOfWeek > 5) {
                dayTo = dayTo.plusDays(8 - dayTo.dayOfWeek)
            }
            this.setStartDate(dayTo.toDate())
            val duration = Days.daysBetween(dateFrom, dateTo).days + 1
            proposedLeave.add(Leave(this.getFirstWorkingDay(days), dayTo.toDate(), duration))
        }

        return proposedLeave
    }
}